lib/pacemaker/pcmk_sched_probes.c

```diff
@@ -178,90 +178,98 @@ add_probe_orderings_for_stops(pe_working_set_t *data_set)
-              probe->node ? probe->node->details->uname: "",
+              ((probe->node == NULL)? "" : probe->node->details->uname),
               after->uuid,
-              after->node ? after->node->details->uname : "");
+              ((after->node == NULL)? "" : after->node->details->uname));
 
-    if (after->rsc
-        /* Better not build a dependency directly with a clone/group.
-         * We are going to proceed through the ordering chain and build
-         * dependencies with its children.
-         */
-        && after->rsc->variant == pe_native
-        && probe->rsc != after->rsc) {
+    /* Add restart orderings if "then" is for a different primitive.
+     * Orderings for collective resources will be added later.
+     */
+    if ((after->rsc != NULL) && (after->rsc->variant == pe_native)
+        && (probe->rsc != after->rsc)) {
 
             GList *then_actions = NULL;
-            enum pe_ordering probe_order_type = pe_order_optional;
 
             if (pcmk__str_eq(after->task, RSC_START, pcmk__str_casei)) {
-                then_actions = pe__resource_actions(after->rsc, NULL, RSC_STOP, FALSE);
+                then_actions = pe__resource_actions(after->rsc, NULL, RSC_STOP,
+                                                    FALSE);
 
             } else if (pcmk__str_eq(after->task, RSC_PROMOTE, pcmk__str_casei)) {
-                then_actions = pe__resource_actions(after->rsc, NULL, RSC_DEMOTE, FALSE);
+                then_actions = pe__resource_actions(after->rsc, NULL,
+                                                    RSC_DEMOTE, FALSE);
             }
 
-            for (gIter = then_actions; gIter != NULL; gIter = gIter->next) {
-                pe_action_t *then = (pe_action_t *) gIter->data;
+            for (iter = then_actions; iter != NULL; iter = iter->next) {
+                pe_action_t *then = (pe_action_t *) iter->data;
 
-                // Skip any pseudo action which for example is implied by fencing
-                if (pcmk_is_set(then->flags, pe_action_pseudo)) {
-                    continue;
+                // Skip pseudo-actions (for example, those implied by fencing)
+                if (!pcmk_is_set(then->flags, pe_action_pseudo)) {
+                    order_actions(probe, then, pe_order_optional);
                 }
-
-                order_actions(probe, then, probe_order_type);
             }
             g_list_free(then_actions);
     }
 
-    if (after->rsc
-        && after->rsc->variant > pe_group) {
+    /* Detect whether "then" is an interleaved clone action. For these, we want
+     * to add orderings only for the relevant instance.
+     */
+    if ((after->rsc != NULL)
+        && (after->rsc->variant > pe_group)) {
         const char *interleave_s = g_hash_table_lookup(after->rsc->meta,
                                                        XML_RSC_ATTR_INTERLEAVE);
 
         interleave = crm_is_true(interleave_s);
-
         if (interleave) {
```

Comment:
It looks like the only point of the interleave variable is this conditional, so you could get rid of it and replace this with `if (crm_is_true(interleave_s)) {`.